### PR TITLE
VB Feature: OptionalParameterDefaults.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
**Prototype** Implementation of VBLang [PR 13](https://github.com/dotnet/vblang/pull/13)
Currently doesn't have unit tests.

@jaredpar Am I targeting the right base branch?

@AnthonyDGreen Is this about the expected size of Pull Request (-unit tests) for should a feature?
I've refactored the existing `optional parameter` code into separate methods, which allowed me to reuse some of them from the method handling the feature.

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb [hunk 1]
```diff
@@ -3043,165 +3043,144 @@ ProduceBoundNode:
...
-                ' is a possible outcome of running an obfuscator over a valid DLL 
-                If defaultConstantValue.IsBad Then
-                    defaultConstantValue = ConstantValue.Null
+                If methodSymbol IsNot Nothing Then
+                    Dim argument = New BoundLiteral(syntax, ConstantValue.Null, param.Type)
+                    defaultArgument = New BoundObjectCreationExpression(syntax, methodSymbol, ImmutableArray.Create(Of BoundExpression)(argument),
+                                                                        Nothing, methodSymbol.ContainingType)
                 End If
+            Else
+                defaultArgument = OptionalParameterWithExplicitDefaultValue(param, syntax, diagnostics, callerInfoOpt, defaultConstantValue)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb]: This line needs to be conditional on the availability of the feature.

**Comment by AnthonyDGreen** [comment]
Churn is looking much smaller. A lot easier to reason about.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot please retest this

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please

**Comment by dpoeschl** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by dpoeschl** [comment]
retest windows_release_vs-integration_prtest please

**Comment by AdamSpeight2008** [comment]
@AnthonyDGreen This is in fairly complete state, just needs a review.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by AdamSpeight2008** [comment]
Assertion Failure, triggered by typing `=` after `As String`
```
Exception thrown: 'System.OperationCanceledException' in mscorlib.dll
---- DEBUG ASSERTION FAILED ----
---- Assert Short Message ----
Unexpected difference at offset 583: Line 18, Column 13 "    End Sub"
---- Assert Long Message ----

   at Microsoft.CodeAnalysis.SyntaxTreeExtensions.VerifySource(SyntaxTree tree, IEnumerable`1 changes)
   at Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxTree.WithChanges(SourceText newText, TextChangeRange[] changes)
   at Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxTree.WithChangedText(SourceText newText)
   at Microsoft.CodeAnalysis.DocumentState.IncrementallyParse(TextAndVersion newTextAndVersion, TreeAndVersion oldTreeAndVersion, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.DocumentState.<IncrementallyParseTreeAsync>d__15.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.CodeAnalysis.DocumentState.IncrementallyParseTreeAsync(ValueSource`1 oldTreeSource, ValueSource`1 newTextSource, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.DocumentState.<>c__DisplayClass14_0.<CreateLazyIncrementallyParsedTree>b__0(CancellationToken c)
   at Roslyn.Utilities.AsyncLazy`1.StartAsynchronousComputation(AsynchronousComputationToStart computationToStart, Request requestToCompleteSynchronously, CancellationToken callerCancellationToken)
   at Roslyn.Utilities.AsyncLazy`1.GetValueAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.DocumentState.<GetSyntaxTreeAsync>d__41.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.CodeAnalysis.DocumentState.GetSyntaxTreeAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Document.GetSyntaxTreeAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Host.BackgroundParser.<>c__DisplayClass20_0.<ParseDocumentAsync>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

Exception thrown: 'System.InvalidOperationException' in Microsoft.CodeAnalysis.Workspaces.dll
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in Microsoft.VisualStudio.LanguageServices.dll

```

**Comment by AdamSpeight2008** [comment]
@jcouv Just the two VS Integration check failing, which seem to be unrelated to the content of this pull request.

**Comment by jcouv** [comment]
VS integration tests are failing on Roslyn.VisualStudio.IntegrationTests.VisualBasic.BasicWinForms.RenameControl ([details](https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/5516/))

FYI @dotnet/roslyn-infrastructure

**Comment by jcouv** [comment]
@AdamSpeight2008 Yes, this test failure seems unrelated to the PR.

**Comment by AdamSpeight2008** [comment]
@jcouv It may have already been fixed I've rebased against `master` in a while.

**Comment by jcouv** [comment]
@AdamSpeight2008 You can leave the PR against `master` for now, but there is a good chance you'll have to rebase after review (branches usually get shuffled when getting close to release, as we are with 15.3).

**Comment by AnthonyDGreen** [comment]
Syncing sources now from your repo, @AdamSpeight2008 
I'm going to bang on it to see if it crashes a lot.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
@AnthonyDGreen 
Any feedback?

**Comment by AdamSpeight2008** [comment]
May I have review please?

**Comment by AdamSpeight2008** [comment]
I think the failures are do to a bad merge / rebase, I'll look at pushing correction Friday.

**Comment by paul1956** [comment]
An interesting test would be to run this codefix on a copy of Roslyn source "Solution" because I believe you will get an OutOfMemory exception. Yet this is a perfect example of where a solution wide CodeFix is needed.

**Comment by AdamSpeight2008** [comment]
@paul1956 It would be a good test. *You'll be able to see the smoke from VS for miles.*

**Comment by AdamSpeight2008** [comment]
Fixed up the issues with bad merge. There are failing unit-test for `Non-TrailingNamedArguments` 
and couple related to this feature.

**Comment by AdamSpeight2008** [comment]
Failure in NamedArgument Tests are due to
`Debug.Assert(argumentNames.IsDefault OrElse (argumentNames.Length > 0 AndAlso argumentNames.Length = arguments.Length)` asserting because
`argumentNames.Length = 2` and `arguments.Length=1` in `CollectOverloadedCandidates`

**Comment by paul1956** [comment]
@AdamSpeight2008 isn't that a problem? If there is a limit of X fixes before bad things happen how about limiting code fixes to a smaller number or 75% of available memory, or fix the algorithm to aggregate each document. I would be much happier if a code fix stopped after 50 (100, 500...) fixes (and I could run it again) which might be one fix in each of 50 files or 50 fixes in one document, then what I have today, where I have to search for the error, go to the file, fix all the issues  then find another file (which might have 1 issue), and repeat for each file. Switching from Implicit to Explicit declarations in just the Roslyn VB CodeAnalysis project is 39,000 changes and going the other way is even more. Changing to Optional Default Parameter is around 6000 changes in ~800 files (very rough guess).

**Comment by AdamSpeight2008** [comment]
I've open a new issue #22591 to discuss this side discussion @paul1956 .

**Comment by AdamSpeight2008** [comment]
@AnthonyDGreen  Can you see why the assertion is failing? 
```vbnet
    Function Goo(<CallerLineNumber> Optional line As Integer) As Integer
        Return line
    End Function
```
Typing a space after `As Integer` causes an assertion failure.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot please retest this

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please

**Comment by AdamSpeight2008** [comment]
```
✘  Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.TestExclusiveNamedParameterCompletion
Assert.NotNull() Failure
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.TestState.VB$StateMachine_35_AssertCompletionSession.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.VB$StateMachine_47_TestExclusiveNamedParameterCompletion.MoveNext() in D:\roslyn-AdamSpeight2008\src\EditorFeatures\Test2\IntelliSense\VisualBasicCompletionCommandHandlerTests.vb:line 1041
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

 Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.TestExclusiveNamedParameterCompletion2
Assert.NotNull() Failure
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.TestState.VB$StateMachine_35_AssertCompletionSession.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.VB$StateMachine_48_TestExclusiveNamedParameterCompletion2.MoveNext() in D:\roslyn-AdamSpeight2008\src\EditorFeatures\Test2\IntelliSense\VisualBasicCompletionCommandHandlerTests.vb:line 1070
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

 Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.TestNonTrailingNamedArgumentInVB15_3
System.NullReferenceException : Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.TestState.VB$StateMachine_40_AssertSelectedCompletionItem.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.VB$StateMachine_121_TestNonTrailingNamedArgumentInVB15_3.MoveNext() in D:\roslyn-AdamSpeight2008\src\EditorFeatures\Test2\IntelliSense\VisualBasicCompletionCommandHandlerTests.vb:line 2546
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

 Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.TestNonTrailingNamedArgumentInVB15_5
System.NullReferenceException : Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.TestState.VB$StateMachine_40_AssertSelectedCompletionItem.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Editor.UnitTests.IntelliSense.VisualBasicCompletionCommandHandlerTests.VB$StateMachine_122_TestNonTrailingNamedArgumentInVB15_5.MoveNext() in D:\roslyn-AdamSpeight2008\src\EditorFeatures\Test2\IntelliSense\VisualBasicCompletionCommandHandlerTests.vb:line 2571
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Comment by AdamSpeight2008** [comment]
The failures seem to stem from NamedParameters. @AnthonyDGreen

**Comment by AdamSpeight2008** [comment]
Closing this in favour of the fresh look at implementation of #24231.

**Pull request closed without merging**